include/native_thread_pool.h

```diff
@@ -0,0 +1,220 @@
+                break;
+            }
+        }
+    }
+    bool is_stopping() const {
+        return stopping.load(std::memory_order_relaxed);
+    }
+    static void pin(unsigned cpu_id) {
+        cpu_set_t cs;
+        CPU_ZERO(&cs);
+        CPU_SET(cpu_id, &cs);
+        [[maybe_unused]] auto r = pthread_setaffinity_np(
+          pthread_self(), sizeof(cs), &cs);
+    }
+    seastar::semaphore& local_free_slots() {
+        return submit_queue.local().free_slots;
+    }
+    ThreadPool(const ThreadPool&) = delete;
+    ThreadPool& operator=(const ThreadPool&) = delete;
+
+public:
+    /**
+     * @param queue_sz the depth of pending queue. before a task is scheduled,
+     *                 it waits in this queue. we will round this number to
+     *                 multiple of the number of cores.
+     * @param n_threads the number of threads in this thread pool.
+     * @param cpu the CPU core to which this thread pool is assigned
+     * @note each @c Task has its own crimson::thread::Condition, which
+     * possesses possesses an fd, so we should keep the size of queue under a
+     * reasonable limit.
+     */
+    ThreadPool(size_t n_threads, size_t queue_size, unsigned cpu_id)
+      : queue_size{queue_size} // round_up_to(queue_sz, seastar::smp::count)}
+      , pending{queue_size}
+      , add_task_sem(n_threads) {
+        for (size_t i = 0; i < n_threads; i++) {
+            threads.emplace_back([this, cpu_id] {
+                pin(cpu_id);
+                loop();
+            });
+        }
+    }
+    ~ThreadPool() {
+        for (auto& thread : threads) {
+            thread.join();
+        }
+    }
+    seastar::future<> start() {
+        auto slots_per_shard = queue_size / seastar::smp::count;
+        return submit_queue.start(slots_per_shard);
+    }
+    seastar::future<> stop() {
+        return submit_queue.stop().then([this] {
+            stopping = true;
+            cond.notify_all();
+        });
+    }
+    template<typename Func, typename... Args>
+    auto submit(Func&& func, Args&&... args) {
+        auto packaged = [func = std::move(func),
+                         args = std::forward_as_tuple(args...)] {
+            return std::apply(std::move(func), std::move(args));
+        };
+        return seastar::with_gate(
+          submit_queue.local().pending_tasks,
+          [packaged = std::move(packaged), this] {
+              return local_free_slots().wait().then(
+                [packaged = std::move(packaged), this] {
+                    auto task = new Task{std::move(packaged)};
+                    auto fut = task->get_future();
+                    pending.push(task);
+                    cond.notify_one();
+                    return fut.finally([task, this] {
+                        local_free_slots().signal();
+                        delete task;
+                    });
+                });
+          });
+    }
+
+    seastar::future<> lock() {
+        return add_task_sem.lock();
```

Comment:
This semaphore introduces dependencies between the different isolates which is not necessary. 
We can just use one thread until we will figure out the way to schedule that (maybe something like boost::asio::strand will do the job).